### PR TITLE
fix: use factory functions for config data and remove plain objects

### LINKED ISSUE #32430: `loadConfig` mutates the `imageConfigDefault` object
### What version of Next.js are you using?

12.0.7

### What version of Node.js are you using?

16.13.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

Inside of `assignDefaults` function we merge and copy `defaultConfig` with the passed in `config`.

```ts
// https://github.com/vercel/next.js/blob/v12.0.7/packages/next/server/config.ts#L14
const result = { ...defaultConfig, ...config }
```

Then later mutates `result` by assigning sub-properties.

```ts
        // https://github.com/vercel/next.js/blob/v12.0.7/packages/next/server/config.ts#L205
        images.domains.push(new URL(config.assetPrefix).hostname)
      
      // https://github.com/vercel/next.js/blob/v12.0.7/packages/next/server/config.ts#L279
      images.loader = 'default'
      
      // https://github.com/vercel/next.js/blob/v12.0.7/packages/next/server/config.ts#L314
      images.path = `${result.basePath}${images.path}`
```

Because the merge and copy is shallow, when we mutate `result.images`, we are actually mutating the original object `imageConfigDefault` exported from https://github.com/vercel/next.js/blob/v12.0.7/packages/next/server/image-config.ts.

This could cause a number of issues, for example, in the case I'm running into, every time `loadConfig` is called the `basePath` property is pre-pended to `imageConfigDefault.path`.

```ts
console.log(imageConfigDefault.path)
// "/_next/image"

await loadConfig("phase-development-server", "./path/to/config/with/foo/base/path");
await loadConfig("phase-development-server", "./path/to/config/with/foo/base/path");
await loadConfig("phase-development-server", "./path/to/config/with/foo/base/path");

console.log(imageConfigDefault.path)
// "/foo/foo/foo/_next/image"
```

See the [codesandbox](https://codesandbox.io/s/immutable-thunder-g3bj8?file=/server.js) for an example. 

### Expected Behavior

Calling `loadConfig` with a config that has a `basePath` should not mutate `imageConfigDefault` or `defaultConfig`.

### To Reproduce

https://codesandbox.io/s/immutable-thunder-g3bj8?file=/server.js

### PR DESCRIPTION
**Pull request by cdierkens**
- fixes #32430 
- closes #32431 
- closes #32474 

See [codesandbox](https://codesandbox.io/s/immutable-thunder-g3bj8?file=/server.js) for reproduction.

## Bug

- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [x] Errors have helpful link attached, see `contributing.md`
- [x] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Review by styfle — DISMISSED** [review]
Thanks!

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 922a6cb8078c0aaefb85ba36b5b20916733a1814

**test/integration/config-experimental-warning/test/index.test.js** 
- Promise in next config > should not show warning with default config from function

<details>
<summary>Expand output</summary>

  ● Promise in next config › should not show warning with default config from function

    expect(received).not.toMatch(expected)

    Expected pattern: not /experimental feature/
    Received string:      "warn  - You have enabled experimental feature(s).
    warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

    warn  - The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config
    "

      19 |       }
      20 |     `)
    > 21 |
         | ^
      22 |     const { stderr } = await nextBuild(appDir, [], { stderr: true })
      23 |     expect(stderr).not.toMatch(/experimental feature/)
      24 |   })

      at Object.<anonymous> (integration/config-experimental-warning/test/index.test.js:21:28)

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| buildDuration | 18.1s | 18.4s | ⚠️ +280ms |
| buildDurationCached | 3.3s | 3.4s | ⚠️ +74ms |
| nodeModulesSize | 350 MB | 350 MB | ⚠️ +958 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.963 | 2.951 | -0.01 |
| / avg req/sec | 843.88 | 847.09 | +3.21 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.326 | 1.314 | -0.01 |
| /error-in-render avg req/sec | 1885.16 | 1903.23 | +18.07 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 30.2 kB | 30.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 74 kB | 74 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.37 kB | 1.37 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.37 kB | 2.37 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 4.73 kB | 4.74 kB | ⚠️ +8 B |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.13 kB | 2.13 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ⚠️ +8 B |

</details>

<details>
<summary><strong>Client Build Manifests</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 460 B | ⚠️ +1 B |
| Overall change | 459 B | 460 B | ⚠️ +1 B |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>_buildManifest.js</strong></summary>

```diff
@@ -12,7 +12,7 @@ self.__BUILD_MANIFEST = {
   ],
   "/head": ["static\u002Fchunks\u002Fpages\u002Fhead-7100d3b2a548f0e4.js"],
   "/hooks": ["static\u002Fchunks\u002Fpages\u002Fhooks-538d621a0e670391.js"],
-  "/image": ["static\u002Fchunks\u002Fpages\u002Fimage-7741d74715a0ad4f.js"],
+  "/image": ["static\u002Fchunks\u002Fpages\u002Fimage-3fb973e65bad1c01.js"],
   "/link": ["static\u002Fchunks\u002Fpages\u002Flink-76af1f5604a3f190.js"],
   "/routerDirect": [
     "static\u002Fchunks\u002Fpages\u002FrouterDirect-76232dd6bc335a24.js"
```
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -202,6 +202,7 @@
         }
         return target;
       }
+      var imageConfigDefault = (0, _imageConfig).getImageConfigDefault();
       var loadedImageURLs = new Set();
       var allImgs = new Map();
       var perfObserver;
@@ -246,7 +247,7 @@
             imageSizes: [16, 32, 48, 64, 96, 128, 256, 384],
             path: "/_next/image",
             loader: "default"
-          } || _imageConfig.imageConfigDefault,
+          } || imageConfigDefault,
         configDeviceSizes = ref.deviceSizes,
         configImageSizes = ref.imageSizes,
         configLoader = ref.loader,
@@ -1091,7 +1092,7 @@
       Object.defineProperty(exports, "__esModule", {
         value: true
       });
-      exports.imageConfigDefault = exports.VALID_LOADERS = void 0;
+      exports.getImageConfigDefault = exports.VALID_LOADERS = void 0;
       const VALID_LOADERS = [
         "default",
         "imgix",
@@ -1100,7 +1101,7 @@
         "custom"
       ];
       exports.VALID_LOADERS = VALID_LOADERS;
-      const imageConfigDefault = {
+      const getImageConfigDefault = () => ({
         deviceSizes: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
         imageSizes: [16, 32, 48, 64, 96, 128, 256, 384],
         path: "/_next/image",
@@ -1109,8 +1110,8 @@
         disableStaticImages: false,
         minimumCacheTTL: 60,
         formats: ["image/webp"]
-      };
-      exports.imageConfigDefault = imageConfigDefault;
+      });
+      exports.getImageConfigDefault = getImageConfigDefault;
 
       //# sourceMappingURL=image-config.js.map
```
</details>
</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| buildDuration | 19.6s | 19.5s | -138ms |
| buildDurationCached | 3.4s | 3.4s | ⚠️ +6ms |
| nodeModulesSize | 350 MB | 350 MB | ⚠️ +958 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.956 | 2.957 | 0 |
| / avg req/sec | 845.82 | 845.53 | ⚠️ -0.29 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.344 | 1.395 | ⚠️ +0.05 |
| /error-in-render avg req/sec | 1860.63 | 1792.18 | ⚠️ -68.45 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 30.3 kB | 30.3 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 74.2 kB | 74.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 906 B | 906 B | ✓ |
| image-HASH.js gzip | 4.75 kB | 4.76 kB | ⚠️ +11 B |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.19 kB | 2.19 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.1 kB | 14.1 kB | ⚠️ +11 B |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | cdierkens/next.js images-config-factory-breaking | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>_buildManifest.js</strong></summary>

```diff
@@ -12,7 +12,7 @@ self.__BUILD_MANIFEST = {
   ],
   "/head": ["static\u002Fchunks\u002Fpages\u002Fhead-7100d3b2a548f0e4.js"],
   "/hooks": ["static\u002Fchunks\u002Fpages\u002Fhooks-538d621a0e670391.js"],
-  "/image": ["static\u002Fchunks\u002Fpages\u002Fimage-7741d74715a0ad4f.js"],
+  "/image": ["static\u002Fchunks\u002Fpages\u002Fimage-3fb973e65bad1c01.js"],
   "/link": ["static\u002Fchunks\u002Fpages\u002Flink-76af1f5604a3f190.js"],
   "/routerDirect": [
     "static\u002Fchunks\u002Fpages\u002FrouterDirect-76232dd6bc335a24.js"
```
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -202,6 +202,7 @@
         }
         return target;
       }
+      var imageConfigDefault = (0, _imageConfig).getImageConfigDefault();
       var loadedImageURLs = new Set();
       var allImgs = new Map();
       var perfObserver;
@@ -246,7 +247,7 @@
             imageSizes: [16, 32, 48, 64, 96, 128, 256, 384],
             path: "/_next/image",
             loader: "default"
-          } || _imageConfig.imageConfigDefault,
+          } || imageConfigDefault,
         configDeviceSizes = ref.deviceSizes,
         configImageSizes = ref.imageSizes,
         configLoader = ref.loader,
@@ -1091,7 +1092,7 @@
       Object.defineProperty(exports, "__esModule", {
         value: true
       });
-      exports.imageConfigDefault = exports.VALID_LOADERS = void 0;
+      exports.getImageConfigDefault = exports.VALID_LOADERS = void 0;
       const VALID_LOADERS = [
         "default",
         "imgix",
@@ -1100,7 +1101,7 @@
         "custom"
       ];
       exports.VALID_LOADERS = VALID_LOADERS;
-      const imageConfigDefault = {
+      const getImageConfigDefault = () => ({
         deviceSizes: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
         imageSizes: [16, 32, 48, 64, 96, 128, 256, 384],
         path: "/_next/image",
@@ -1109,8 +1110,8 @@
         disableStaticImages: false,
         minimumCacheTTL: 60,
         formats: ["image/webp"]
-      };
-      exports.imageConfigDefault = imageConfigDefault;
+      });
+      exports.getImageConfigDefault = getImageConfigDefault;
 
       //# sourceMappingURL=image-config.js.map
```
</details>
</details>
Commit: 922a6cb8078c0aaefb85ba36b5b20916733a1814

**Comment by cdierkens** [comment]
Sorry for the noise. I was getting this branch up to date with `canary` and my local `canary` branch wasn't up to date with the `next.js` one.

File: packages/next/server/config.ts [hunk 1]
```diff
@@ -643,7 +652,7 @@ export default async function loadConfig(
     }
   }
 
-  const completeConfig = defaultConfig as NextConfigComplete
+  const completeConfig = getDefaultConfig() as NextConfigComplete
```
Comment (styfle) [inline packages/next/server/config.ts:655]:
```suggestion
  const completeConfig = defaultConfig as NextConfigComplete
```

Comment (cdierkens) [reply packages/next/server/config.ts:655]: `defaultConfig` is instantiated in `assignDefaults` and this line is in `loadConfig`. 

Moving it to the static scope would somewhat defeat the purpose of the factory function as well, as that opens the ability to mutate within this module scope.

Comment (styfle) [reply packages/next/server/config.ts:655]: Let's pass it into `assignDefaults()` then?

**Pull request closed without merging**